### ISSUE TITLE
Bugged outputs with KV Cache

### ISSUE THREAD

**Issue by JLenzy** [lucidrains/x-transformers#322]
I'm trying to incorporate KV cache into my inference pipeline, using TransformerWrapper (but not Autoregressive Wrapper). 

I define my model as follows (the specifics are loaded from a config):

```
self.decoder = TransformerWrapper(
            num_tokens=self.vocab_size,
            max_seq_len=self.max_len,
            tie_embedding=self.tie_embedding,
            emb_dropout=self.embedding_dropout,
            attn_layers=Decoder(
                dim=self.d_model,
                depth=self.n_layers,
                heads=self.n_heads,
                weight_tie_layers=self.weight_tie_layers,
                use_rmsnorm=self.use_rmsnorm,
                ff_glu=self.ff_glu,
                ff_no_bias=self.ff_no_bias,
                layer_dropout=self.layer_dropout,  # stochastic depth - dropout entire layer
                attn_dropout=self.attn_dropout,
                ff_dropout=self.ff_dropout,
                rotary_pos_emb=True
            )
        )
```

and the forward method:
```
def forward(
        self,
        tgt: Tensor,
        tgt_key_padding_mask: Optional[Tensor] = None,
        cache: Optional[Tensor] = None,
        return_hiddens: bool = False
    ):
        return self.decoder(tgt, mask=tgt_key_padding_mask, return_intermediates=return_hiddens, cache=cache)
```

Below is an excerpt of my generation code. What I have noticed is that generation 'works' with this version, but the outputs are really.....bad in quality. If I uncomment the `cache=None` (thus making it recalculate the full sequence each time), there is a huge jump in quality. Any idea what is going on here?

```
cache = None
        
        with tqdm(total=max_tokens, initial=pred_tokens.shape[-1]) as pbar:
            for idx, pos in enumerate(range(pred_tokens.shape[-1], max_tokens)):
                # For the first iteration, process the entire input
                #cache = None
                if cache is None:
                    output, cache = self.forward(pred_tokens, cache=None, return_hiddens=True)
                else:
                    # For subsequent iterations, only process the last token
                    # This is a major optimization that speeds up inference
                    last_token = pred_tokens[:, -1:]
                    output, cache = self.forward(last_token, cache=cache, return_hiddens=True)
                
                scaled_logits = output[:, -1, :] / temperature
                probs = torch.softmax(scaled_logits, dim=-1)

                idx_next = sample_next_logit(
                    logits=probs,
                    top_p=top_p,
                    top_k=top_k,
                    sampling_type=sample_type
                )
                pred_tokens = torch.cat((pred_tokens, idx_next), dim=1)
```

**Comment by lucidrains** [comment]
@JLenzy hmm, does it work if you use autoregressive wrapper?

**Comment by lucidrains** [comment]
@JLenzy ohh, at inference time, are you setting it to eval mode? notice you have stochastic depth turned on

**Comment by JLenzy** [comment]
@lucidrains I think I found the issue. Claude had suggested trimming the input to just the newest tokens (see if 'if' statement at the beginning of inference loop above), but that was messing it up. I just had to simplify to:

```
with tqdm(total=max_tokens, initial=pred_tokens.shape[-1]) as pbar:
            for idx, pos in enumerate(range(pred_tokens.shape[-1], max_tokens)):
                
                # Forward pass with KV cache
                output, cache = self.forward(pred_tokens, cache=cache, return_hiddens=True)
```

So it takes the full 'pred_tokens' input at each forward pass. 

(also - to answer your question above, stochastic depth is actually 'off' for this model, so not a concern here).

**Comment by lucidrains** [comment]
@JLenzy i've personally am finding the new gemini pro 2.5 to be much better than Claude

give it a try and let me know how it goes!

**Comment by lucidrains** [comment]
@JLenzy also, do let me know if the autoregressive wrapper is lacking in any way
